lib/Sema/CSSimplify.cpp

```diff
@@ -6161,8 +6161,10 @@ bool ConstraintSystem::repairFailures(
     if (lhs->isAny()) {
       rhs.visit([&](Type type) {
         if (auto *typeVar = type->getAs<TypeVariableType>()) {
-          assignFixedType(typeVar,
-                          PlaceholderType::get(getASTContext(), typeVar));
+          if (!getFixedType(typeVar)) {
+            assignFixedType(typeVar,
+                            PlaceholderType::get(getASTContext(), typeVar));
+          }
```

Comment:
Would it be better to `simplifyType(rhs)` here? Mainly asking as this pattern is used in a couple of other places:

https://github.com/apple/swift/blob/1c3667b52aa7888f7e2cc253b725a7ed055e4ef2/lib/Sema/CSStep.cpp#L1106-L1130

https://github.com/apple/swift/blob/1c3667b52aa7888f7e2cc253b725a7ed055e4ef2/lib/Sema/BuilderTransform.cpp#L1178-L1182

And I've just realized there's a very subtle footgun to their implementation: because `getAs` looks through sugar, if it visits e.g a ParenType, then it would end up trying to `assignFixedType` twice, which would blow up. Maybe we should factor out this logic into a function that ensures this doesn't happen?

Reply:
`simplifyType` transforms which I am generally cautious about, that’s why I didn’t use it here, and in general we shouldn’t be using `assignFixedType` directly too. I considered to use Bind constraint instead but my worry is that the flags are not always reliable to indicate whether new constraints are allowed…

Reply:
Yeah, I'm just wondering if there's something we can factor out to avoid the footgun with `getAs`. Maybe we should at least change them to use `dyn_cast` for now to avoid potential issues (I haven't yet tried to come up with cases that would hit the issue, but it seems like it should be possible).

Reply:
`getFixedType` is the only way to go I guess. I'm going to extract this into a function.

Reply:
Alright, done!

Reply:
Thanks!